book/content/testing/01/good_practice.md

```diff
@@ -0,0 +1,309 @@
+
+Consider setting up continuous integration (discussed in the continuous integration chapter) on your project. 
+This will automatically run your tests each time you make a change to your code and, depending on the continuous integration software you use, will notify you if any of the tests fail.
+
+<a name="Consider_how_long_it_takes_your_tests_to_run"></a>
+### Consider how long it takes your tests to run
+
+Some tests, like [unit tests](#Unit_tests) only test a small piece of code and so typically are very fast. 
+However other kinds of tests, such as [system tests](#System_tests) which test the entire code from end to end, may take a long time to run depending on the code. 
+As such it can be obstructive to run the entire test suite after each little bit of work. 
+In that case it is better to run lighter weight tests such as unit tests frequently, and longer tests only once per day overnight. 
+It is also good to scale the number of each kind of tests you have in relation to how long they take to run. 
+You should have a lot of unit tests (or other types of tests that are fast) but much fewer tests which take a long time to run.
+
+<a name="Document_the_tests_and_how_to_run_them"></a>
+### Document the tests and how to run them
+
+It is important to provide documentation that describes how to run the tests, both for yourself in case you come back to a project in the future, and for anyone else that may wish to build upon or reproduce your work. 
+This documentation should also cover subjects such as
+
+- Any resources e.g. test dataset files that are required
+- Any configuration/settings adjustments needed to run the tests
+- What software (such as [testing frameworks](#Use_a_testing_framework)) need to be installed
+
+Ideally, you would provide scripts to set up and configure any resources that are needed.
+
+<a name="Test_realistic_cases"></a>
+### Test realistic cases
+
+Make the cases you test as realistic as possible. 
+If for example, you have dummy data to run tests on you should make sure that data is as similar as possible to the actual data. 
+If your actual data is messy with a lot of null values, so should your test dataset be.
+
+<a name="Use_a_testing_framework"></a>
+### Use a testing framework
+
+There are tools available to make writing and running tests easier, these are known as testing frameworks. 
+Find one you like, learn about the features it offers, and make use of them. Common testing frameworks (and the languages they apply to) include:
+
+- Language agnostic
+  - CTest, test runner for executables, bash scripts, etc, great for legacy code hardening
+- C++
+  - Catch
+  - CppTest
+  - Boost::Test
+  - google-test      
+- C
+  - all C++ frameworks
+  - Check
+  - CUnit
+- Python
+  - pytest (recommended)
+  - unittest comes with standard Python library
+- R unit-tests
+  - RUnit
+  - svUnit (works with SciViews GUI)
```

Comment:
I would add `testthat` and `tinytest` to the list and remove `RUnit` or say it is not recommended. 

`testthat` is by far the most popular package for R (actually the example of asserting an equality below is done with `testthat`) and is actively developed. `tinytest` is also being actively developed whereas `RUnit`'s last release is more than a year old, which triggered some packages have stopped using it (i.e. [RcppArmadillo](https://github.com/RcppCore/RcppArmadillo)).

I was thinking on adding a simple PR for this alone, but I don't want to step on your toes nor difficult this amazing refactoring, so I am fine with you adding the changes directly here if it makes your life easier.

Reply:
Thanks @oriolabril!

I think this pull request is a little stale (@rosiehigman has a new job!) so I'm not sure who is going to pick this up, so if you wanted to do a little PR then please go for it, but otherwise just leave this comment here so someone can find it again in the future!

@annakrystalli - just tagging you in case you have strong opinions about R testing (I don't know anything John Snow!) If

Reply:
I opened #760 to change these 2 lines